### PR TITLE
workspace-switcher: support i3 WM

### PR DESCRIPTION
**Pull request by raveit65**
Fixes https://github.com/mate-desktop/mate-panel/issues/711

### CODE REVIEW

File: applets/wncklet/workspace-switcher.c [hunk 1]
```diff
@@ -147,8 +160,10 @@ static void window_manager_changed(WnckScreen* screen, PagerData* pager)
 
 	if (!wm_name)
 		pager->wm = PAGER_WM_UNKNOWN;
-	else if (strcmp(wm_name, "Metacity (Marco)") == 0)
+	else if (strcmp(wm_name, "Marco") == 0)
```
Comment (muktupavels) [inline applets/wncklet/workspace-switcher.c]: Did you test this? Seems that `marco` still used `Metacity (Marco)` as wm name:
https://github.com/mate-desktop/marco/blob/master/src/core/display.c#L667

Comment (raveit65) [reply applets/wncklet/workspace-switcher.c]: No, i will do, otherwise i will revert this change.

**Comment by raveit65** [comment]
I have no idea if this will work with i3.

**Comment by lukefromdc** [comment]
I do not have i3 installed, and have never worked in tiling WMs for that matter

**Comment by raveit65** [comment]
@muktupavels 
Thanks a lot for helping

**Comment by raveit65** [comment]
@lukefromdc 
Major improvement is that setting dialog from workspace-switcher is now working with metacity.
I forgot to mention that here.

**Comment by lukefromdc** [comment]
I am really not that familiar with this, as I have always used the miniature window layouts and never workspace names in the switcher, and normally use compiz. Thus realy not that familiar with the normal behavior of this at all.

**Comment by raveit65** [comment]
You never used this dialog?
workspace-switcher-dialog

**Comment by lukefromdc** [comment]
I've tested it but only actually used the top portion, as I'm almost always running compiz and never name workspaces

**Comment by anton-z-s** [comment]
I'll test it in 2 days.

**Comment by raveit65** [comment]
> I tried installing and running i3 and all I got was a caja desktop with nothing on it - couldn't even open a terminal... The i3 thing was not related to this patch, I just couldn't figure out how to use it.

I had exactly the same problem :)

**Comment by anton-z-s** [comment]
@vkareh caja is overlapping all the panels and apps. Remove _filemanager_ from org > mate > desktop > session > required-component.

**Comment by raveit65** [comment]
> caja is overlapping all the panels and apps. Remove filemanager from org > mate > desktop > session > required-component.

Ok this working, but i have no idea to setup workspace names and number of workspaces in i3.
The workspace-switcher dialog doesn't show this options.

**Comment by anton-z-s** [comment]
in ~/.config/i3/config replace
`bindsym $mod+1 workspace 1`
with 
`bindsym $mod+1 workspace 1: mail` 
this'll work if you answered _Yes, generate .i3/config_   on first run.

**Comment by raveit65** [comment]
I will cherry-pick the commit for metacity WM to master and 1.20, as long we wait for testing i3 WM.

**Comment by anton-z-s** [comment]
Ok, after installing it with `make install` and re-logging i see no changes. I've expected option "Show workspace names in switcher" to appear, but it's still missing.
I've doubt i use compiled version of mate panel. But it's installed: 
```
# apt list mate-panel -a
Listing... Done
mate-panel/bionic 1.20.0-3ubuntu1 amd64 [upgradable from: 1.20.0-2ubuntu1]
mate-panel/now 1.20.0-2ubuntu1 amd64 [installed,upgradable to: 1.20.0-3ubuntu1]
```
What am i doing wrong?

**Comment by raveit65** [comment]
> Ok, after installing it with make install and re-logging i see no changes.

I am using RPMS, deb package-system is far away from my knowledge.
I think you need to un-install the normal deb mate-panel package and
./configure --prefix=/usr
make
sudo make install

Maybe someone knows to rebuild your existing deb mate-panel package with the patch?
This should be the better solution.
@lukefromdc @muktupavels @monsta @flexiondotorg @sc0w

**Comment by raveit65** [comment]
Btw. i am i doubt that you will ever see an option in workspace-switcher editor as it use dconf/gsettings
Does i3 use dconf/gsettings?

**Comment by raveit65** [comment]
Opps, which distro are you using?
 deb based or rpm based?

**Comment by raveit65** [comment]
Ok, i see you're using f27
i will create a rpm for you.

**Comment by raveit65** [comment]
@anton-z-s 
Here we go https://koji.fedoraproject.org/koji/taskinfo?taskID=25929781
Follow links for your arch, download rpms and update existing packages.

**Comment by anton-z-s** [comment]
@raveit65 i3 doesn't use dconf.
Thanks for the rpm, it works for me, no glitches.

**Comment by muktupavels** [comment]
Post output of `wmctrl -m`.

**Comment by anton-z-s** [comment]
Name: i3
Class: N/A
PID: N/A
Window manager's "showing the desktop" mode: N/A

**Comment by muktupavels** [comment]
The problem is that i3 is not detected. Code checks for `I3` not `i3`...

**Comment by raveit65** [comment]
OK, i will update patch and rpm, I3 --> i3

**Comment by anton-z-s** [comment]
Yeah, i've tried 'I3', that way WM does not even start...
By the way, it check 'Marco', but default value in dconf is 'marco'. So i thought `wnck_screen_get_window_manager_name` capitalized it's output.

**Comment by raveit65** [comment]
new scratch build is on the run
https://koji.fedoraproject.org/koji/taskinfo?taskID=25934746
I changed
```
+	else if (strcmp(wm_name, "I3") == 0)
to
+	else if (strcmp(wm_name, "i3") == 0)
```

This way mate-session starts with i3

[rave@f26 ~]$ gsettings get org.mate.session.required-components windowmanager
'i3'

**Comment by raveit65** [comment]
Hmm, still wondering why this shouldn't work.
```
static void window_manager_changed(WnckScreen* screen, PagerData* pager)
{
	const char *wm_name;

	wm_name = wnck_screen_get_window_manager_name(screen);

	if (!wm_name)
		pager->wm = PAGER_WM_UNKNOWN;
	else if (strcmp(wm_name, "Metacity (Marco)") == 0)
		pager->wm = PAGER_WM_MARCO;
	else if (strcmp(wm_name, "Metacity") == 0)
		pager->wm = PAGER_WM_METACITY;
	else if (strcmp(wm_name, "i3") == 0)
		pager->wm = PAGER_WM_I3;
	else if (strcmp(wm_name, "Compiz") == 0)
		pager->wm = PAGER_WM_COMPIZ;
	else
		pager->wm = PAGER_WM_UNKNOWN;

	update_properties_for_wm(pager);
	pager_update(pager);
}
```
And i see no other check for 'i3' in code.

**Comment by raveit65** [comment]
I rebased PR, but i have no idea to get this working.

**Comment by raveit65** [comment]
Hmm, it works for me with updated commit from https://koji.fedoraproject.org/koji/taskinfo?taskID=25934746
i3-workspace-switcher
And the toggling workspace names with the checkbox works fine.
Only number of workspaces spinbutton doesn't have any function.
I guess i need to disable the spinbutton in UI for i3 WM.
Are you shure you've installed to correct scratch build?
You have to downgrade to normal package versions first before you update to new scratch build.

**Comment by anton-z-s** [comment]
After using `dnf reinstall` instead of `dnf install` it does work. My bad.
Checkbox is here and names are showing, thanks for amazing work!

**Comment by raveit65** [comment]
@anton-z-s 
How can i enable more than one workspace for i3?
I want to test it before i disable non working stuff in UI.

**Comment by anton-z-s** [comment]
Just press _Super+2_ to switch to second workspace.

**Comment by raveit65** [comment]
And how can i restart i3 from command line?

**Comment by anton-z-s** [comment]
It's _Super+Shift+r_

**Comment by raveit65** [comment]
PR is rebased and new scratch build is on the run
https://koji.fedoraproject.org/koji/taskinfo?taskID=25963171

**Comment by anton-z-s** [comment]
That works exactly as expected.

**Pull request merged**